### PR TITLE
Test more webrtc connectivity scenarios

### PR DESCRIPTION
**Pull request by jessetane**
I discovered this project while trying to troubleshoot a specific WebRTC connectivity issue - host to host behind the same NAT. Currently testrtc only appears to test WebRTC connectivity through a relay, leaving out the server reflexive and host scenarios. This pull request adds tests for these scenarios, and reduces possible confusion by splitting out the network capability tests and throughput tests into their own suites.

See #125 for prior relevant discussion.

cc @KaptenJansson @andresusanopinto

### CODE REVIEW

**Comment by jessetane** [comment]
The automated tests are failing in Firefox. From what I can tell, the FF build being used by Travis does not trigger the onerror callback for getUserMedia - I think this prevents the dialog from ever showing the "continue without audio/or video" button and so the tests just hang...

**Comment by jessetane** [comment]
nvm, that's just what happens in FF on my machine... tried on another box and saw the real problem: e7217af

File: src/js/conntest.js [hunk 1]
```diff
@@ -7,100 +7,64 @@
...
-      if (uri.indexOf(transport) !== -1) {
-        newUrls.push(uri);
-      } else if (uri.indexOf('?transport=') === -1 && uri.startsWith('turn')) {
-        newUrls.push(uri + '?' + transport);
+function reflexiveConnectivityTest() {
+  runConnectivityTest(Call.isReflexive, {
+    iceServers: [
+      {
+        urls: [
+          'stun:23.21.150.121'
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Could you use stun.l.google.com:19302 instead?

File: src/js/conntest.js [hunk 2]
```diff
@@ -7,100 +7,64 @@
...
-      reportFatal);
-}
-
-// Test whether it is IPv6 enabled (TODO: test IPv6 to a destination).
-// Turn on IPv6, and try to get an IPv6 host candidate.
-function hasIpv6CandidatesTest() {
-  var params = {optional: [{googIPv6: true}]};
-  gatherCandidates(null, params, Call.isIpv6);
+    runConnectivityTest.bind(null, Call.isRelay),
+    reportFatal);
```
Comment (KaptenJansson) [inline src/js/conntest.js]: this can be on the same line

File: src/js/nettest.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
+/*
+ *  Copyright (c) 2014 The WebRTC project authors. All Rights Reserved.
```
Comment (KaptenJansson) [inline src/js/nettest.js]: Update to 2015.

File: src/js/nettest.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      pc.close();
+      reportFatal('Failed to gather specified candidates');
+    }
+  };
+
+  // Create an audio-only, recvonly offer, and setLD with it.
+  // This will trigger candidate gathering.
+  var createOfferParams = {mandatory: {OfferToReceiveAudio: true}};
+  pc.createOffer(function(offer) { pc.setLocalDescription(offer, noop, noop) ;},
+                 noop, createOfferParams);
```
Comment (KaptenJansson) [inline src/js/nettest.js]: Why do you need a noop function here?

File: src/js/conntest.js [hunk 5]
```diff
@@ -7,100 +7,64 @@
...
-        gatherCandidates(config, null, Call.isRelay);
-      },
-      reportFatal);
-}
+var timeout = null;
 
-// Test whether it can connect via TCP to a TURN server
-// Get a TURN config, and try to get a relay candidate using TCP.
-function tcpConnectivityTest() {
+function relayConnectivityTest() {
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Add a short comment explaining this test.

File: src/js/conntest.js [hunk 6]
```diff
@@ -7,100 +7,64 @@
...
-  for (var i = 0; i < config.iceServers.length; ++i) {
-    var iceServer = config.iceServers[i];
-    var newUrls = [];
-    for (var j = 0; j < iceServer.urls.length; ++j) {
-      var uri = iceServer.urls[j];
-      if (uri.indexOf(transport) !== -1) {
-        newUrls.push(uri);
-      } else if (uri.indexOf('?transport=') === -1 && uri.startsWith('turn')) {
-        newUrls.push(uri + '?' + transport);
+function reflexiveConnectivityTest() {
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Add a short comment explaining this test.

File: src/js/conntest.js [hunk 7]
```diff
@@ -7,100 +7,64 @@
...
-    return reportFatal('Fail to create peer connection: ' + error);
-  }
-
-  // In our candidate callback, stop if we get a candidate that passes |isGood|.
-  pc.onicecandidate = function(e) {
-    // Once we've decided, ignore future callbacks.
-    if (pc.signalingState === 'closed') {
-      return;
-    }
+function hostConnectivityTest() {
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Add a short comment explaining this test.

File: src/js/conntest.js (hunk 2 above)
Comment (jessetane) [inline src/js/conntest.js]: I would agree but the linter gets fussy

File: src/js/nettest.js (hunk 4 above)
Comment (jessetane) [inline src/js/nettest.js]: This file is just a straight copy of what used to be the connectivity test so I didn't write this - however I guess the noop might be needed for Firefox, which seems to get angry if you don't pass onsuccess / onerror callbacks.

Comment (KaptenJansson) [inline src/js/nettest.js]: Aha sorry, I missed that, then it's fine.

File: src/js/conntest.js (hunk 2 above)
Comment (KaptenJansson) [inline src/js/conntest.js]: But it does not care here https://github.com/webrtc/testrtc/pull/127/files#diff-4d6dd141efeb8b75cad3c691015422f8R91

Comment (jessetane) [inline src/js/conntest.js]: The error I get says `Line must be at most 80 characters at src/js/conntest.js :`

Comment (KaptenJansson) [inline src/js/conntest.js]: Ok, lets leave it as is then ;).

**Comment by KaptenJansson** [comment]
@andresusanopinto do you have anything else you want to add?

File: src/js/conntest.js (hunk 2 above)
Comment (jessetane) [inline src/js/conntest.js]: I guess I can try to get it on two lines though instead of three, 1 sec

File: src/js/conntest.js [hunk 8]
```diff
@@ -7,100 +7,72 @@
...
-      var uri = iceServer.urls[j];
-      if (uri.indexOf(transport) !== -1) {
-        newUrls.push(uri);
-      } else if (uri.indexOf('?transport=') === -1 && uri.startsWith('turn')) {
-        newUrls.push(uri + '?' + transport);
+// Set up a datachannel between two peers through a public IP address
+// and verify data can be transmitted and received
+// (packets should stay on the link if behind a router doing NAT)
+function reflexiveConnectivityTest() {
+  runConnectivityTest(Call.isReflexive, {
```
Comment (andresusanopinto) [inline src/js/conntest.js]: We don't want to hard-code a stun server here. This should still use the stun or turn server provided by Call.asyncCreateTurnConfig.

Comment (KaptenJansson) [inline src/js/conntest.js]: Right, that makes sense.

Comment (jessetane) [inline src/js/conntest.js]: Ok, let me know what you think of 9f15cd7

Comment (KaptenJansson) [inline src/js/conntest.js]: I think we should use the turn servers provided by https://github.com/webrtc/testrtc/blob/master/src/js/call.js#L160. Do note that you will get a list of  TURN servers, however they do STUN as well, just change the URL to stun when stun is required:
`turn:23.251.129.26:3479 > stun:23.251.129.26:3479`

Comment (jessetane) [inline src/js/conntest.js]: Ok try c49cd652043ebe2a04f4256bde93b3a6f535c363

Comment (KaptenJansson) [inline src/js/conntest.js]: That looks good! Add it to this PR and then lets merge ;)

**Comment by jessetane** [comment]
@KaptenJansson it's already here 👆

**Comment by KaptenJansson** [comment]
Doh, I looked for the matching commit string but obviously missed it ;).

LGTM!

Thanks for your patience and contribution!

**Pull request merged**